src/common/auth.rs

```diff
@@ -0,0 +1,53 @@
+use super::strings::ct_eq;
+use crate::settings::ServiceConfig;
+
+/// The API keys used for auth
+#[derive(Clone, Debug)]
+pub struct AuthKeys {
+    /// A key allowing Read or Write operations
+    read_write: Option<String>,
+
+    /// A key allowing Read operations
+    read_only: Option<String>,
+}
+
+impl AuthKeys {
+    /// Defines the auth scheme given the service config
+    ///
+    /// Returns None if no scheme is specified.
+    pub fn try_create(service_config: &ServiceConfig) -> Option<Self> {
+        match (
+            service_config.api_key.clone(),
+            service_config.read_only_api_key.clone(),
+        ) {
+            (None, None) => None,
+            (read_write, read_only) => Some(Self {
+                read_write,
+                read_only,
+            }),
+        }
+    }
+
+    /// Check if a key is allowed to read
+    #[inline]
+    pub fn can_read(&self, key: &str) -> bool {
+        self.read_write
+            .as_ref()
+            .map(|rw_key| ct_eq(rw_key, key))
+            .unwrap_or_default()
+            || self
+                .read_only
+                .as_ref()
+                .map(|ro_key| ct_eq(ro_key, key))
+                .unwrap_or_default()
```

Comment:
Micro-optimization: When checking the keys we are always checking `can_write()` first, so it makes sense that `can_read()` checks the read-only key first, so that we don't check the write key twice before checking the read one.

```suggestion
        self.read_only
          .as_ref()
          .map(|ro_key| ct_eq(ro_key, key))
          .unwrap_or_else(|| self.can_write(key))
```